### PR TITLE
Process stack overflow panic elegantly

### LINKED ISSUE #1434: Query with 100 OR conditions overflows stack
**Describe the bug**
`SELECT * FROM table WHERE <condition0> OR <condition1> OR ...` succeeds with 50 conditions but overflows stack with 100 conditions.

**To Reproduce**
Adding this test to arrow-datafusion/datafusion/tests/sql.rs demonstrates the issue:

```
#[tokio::test]
async fn query_with_many_conditions() -> Result<()> {
    let schema = Arc::new(Schema::new(vec![Field::new("c1", DataType::Utf8, true)]));
    let data = RecordBatch::try_new(
        schema.clone(),
        vec![Arc::new(StringArray::from(vec!["foo", "bar"]))],
    )
    .unwrap();

    let table = MemTable::try_new(schema, vec![vec![data]])?;

    let mut ctx = ExecutionContext::new();
    ctx.register_table("test", Arc::new(table)).unwrap();

    let num_conditions = 50;
    println!("attempted a query with {:?} conditions...", num_conditions);
    let where_clause = (0..num_conditions)
        .map(|i| format!("c1 = 'value{:?}'", i))
        .collect::<Vec<String>>()
        .join(" OR ");
    let sql = format!("SELECT * from test where {};", where_clause);
    execute_to_batches(&mut ctx, &sql).await;
    println!("query succeeded with {:?} conditions", num_conditions);

    let num_conditions = 100;
    println!("attempted a query with {:?} conditions...", num_conditions);
    let where_clause = (0..num_conditions)
        .map(|i| format!("c1 = 'value{:?}'", i))
        .collect::<Vec<String>>()
        .join(" OR ");
    let sql = format!("SELECT * from test where {};", where_clause);
    execute_to_batches(&mut ctx, &sql).await;
    println!("query succeeded with {:?} conditions", num_conditions);
    Ok(())
}
```

Output:
```
running 1 test
attempted a query with 50 conditions...
query succeeded with 50 conditions
attempted a query with 100 conditions...

thread 'query_with_many_conditions' has overflowed its stack
fatal runtime error: stack overflow
error: test failed, to rerun pass '--test sql'
```

**Expected behavior**
Expected query to succeed.

### PR DESCRIPTION
**Pull request by xudong963**
# Which issue does this PR close?

Closes #1434 

 # Rationale for this change

Use `SafeRecursion` to support limited recursion to avoid stack overflow.  `maybe_grow`supports unconditionally grow the stack space, by spilling over to the heap if the stack has hit its limit.

# What changes are included in this PR?

Converts a panic! into an Error in some (very important)
Allows the user to set a higher stack limit (but overflows are still possible)

# Are there any user-facing changes?

No

### CODE REVIEW

**Comment by xudong963** [comment]
I think there are some places that need to be wrapped by `maybe_grow`, such as `match expr` in the physical plan, but I want to support them in the next ticket. In this ticket, the skeleton is laid.

PTAL @alamb @houqp @Dandandan

File: datafusion/src/safe_stack.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            return Err(DataFusionError::RecursionLimitErr(self.limit));
+        }
+        *depth += 1;
+        Ok(())
+    }
+}
+
+/// Bytes available in the current stack
+pub const STACKER_RED_ZONE: usize = {
+    64 << 10 // 64KB
```
Comment (xudong963) [inline datafusion/src/safe_stack.rs:55]: This constant is what I feel, I want to know your thoughts.

File: datafusion/src/safe_stack.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+};
+
+/// Allocate a new stack of at least stack_size bytes.
+pub const STACKER_SIZE: usize = {
+    4 << 20 // 4MB
```
Comment (xudong963) [inline datafusion/src/safe_stack.rs:60]: ditto

File: datafusion/src/sql/planner.rs [hunk 3]
```diff
@@ -200,7 +211,9 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
 
     /// Generate a logic plan from an SQL query
     pub fn query_to_plan(&self, query: &Query) -> Result<LogicalPlan> {
-        self.query_to_plan_with_alias(query, None, &mut HashMap::new())
+        self.safe_recursion(|| {
```
Comment (xudong963) [inline datafusion/src/sql/planner.rs:214]: Determine if it is safe to call this function without stack overflow, if it's not safe, return `Err(Recursion exceeded limit)`

File: datafusion/src/sql/planner.rs [hunk 4]
```diff
@@ -113,7 +121,10 @@ fn plan_indexed(expr: Expr, mut keys: Vec<Value>) -> Expr {
 impl<'a, S: ContextProvider> SqlToRel<'a, S> {
     /// Create a new query planner
     pub fn new(schema_provider: &'a S) -> Self {
-        SqlToRel { schema_provider }
+        SqlToRel {
+            schema_provider,
+            project_recursion: ProtectRecursion::new_with_limit(2048),
```
Comment (xudong963) [inline datafusion/src/sql/planner.rs:126]: ditto

**Comment by Jimexist** [comment]
I wonder if it makes sense to just limit number of or statements instead?

**Comment by xudong963** [comment]
> I wonder if it makes sense to just limit number of or statements instead?

First, `Statement` represents the SQL AST node, I don't think it's reasonable to limit the depth of recursion stack by limiting the number of `Statement`.

Second, in some cases, it's preferable to unconditionally grow the stack than returning err or panic.

**Review by alamb — COMMENTED** [review]
I ran out of time today to look at this issue carefully, but will do so tomorrow

**Review by houqp — APPROVED** [review]
This is an acceptable short term workaround, but I think it would be more efficient and more elegant if we rewrite these recursive procedures into iterative procedures.

**Comment by xudong963** [comment]
> This is an acceptable short term workaround, but I think it would be more efficient and more elegant if we rewrite these recursive procedures into iterative procedures.

Do you mean to use push-based model？

**Review by alamb — COMMENTED** [review]
I previously fixed a similar issue like this in:
https://github.com/apache/arrow-datafusion/pull/1047

For the proposed approach (grow the stack if needed), I think this PR is quite elegant -- nicely done @xudong963

However, I think this PR doesn't actually fix the stack overflow, rather it:
1. Converts a `panic!` into an `Error` in some (very important)
2. Allows the user to set a higher stack limit (but overflows are still possible)

The idea of datafusion protecting itself from pathological input cases is 👍 . The thing I worry about with the approach in this PR is that it relies on us annotating all places in the code (both that currently exist as well as may be added in the future) that do recursive walks of the tree with `maybe_grow`). It seems almost inevitable that we will end up missing some and I do think this code will likely add a non trivial overhead to datafusion)

> I wonder if it makes sense to just limit number of or statements instead?

I think the idea from @Jimexist was to limit the depth of any `Expr` tree by running a checking function like

```rust
pub fn check_depth(plan: LogicalPlan) {
  // recursively check all Exprs in all LogicalPlans 
  // for depth greater than 50 (or something)
}
```

I think @houqp's plan of "rewriting to not be recursive" is the best -- I will add a comment describing it shortly

**Comment by alamb** [comment]
> > This is an acceptable short term workaround, but I think it would be more efficient and more elegant if we rewrite these recursive procedures into iterative procedures.

> Do you mean to use push-based model？


@xudong963 , I think what @houqp  is suggesting is to rewrite code that is recursive to not be recursive.

The pattern for Datafusion probably looks like taking code like

```rust
fn visit(expr: &expr)  {
  for child in expr.children() {
    visit(child)
  }
  // do actual expr logic
}
```

And changing it so the state is tracked with a structure on the heap rather than a stack. I think [`VecDeque`](https://doc.rust-lang.org/stable/std/collections/struct.VecDeque.html) is a good one for rust:

```rust
fn visit(expr: &expr)  {
  let mut worklist = VecDequeue::new();
  worklist.push_back(expr);
  while !worklist.is_empty() {
    for child in expr.children() {
      worklist.push_back(child)
    }
    // do actual expr logic
  }
}
```

(aka avoid the call back to `visit`)

**Comment by xudong963** [comment]
> * Converts a `panic!` into an `Error` in some (very important)
> * Allows the user to set a higher stack limit (but overflows are still possible)

Very nice conclusion, You said what I was thinking inside👍



> The thing I worry about with the approach in this PR is that it relies on us annotating all places in the code (both that currently exist as well as may be added in the future) that do recursive walks of the tree with `maybe_grow`). It seems almost inevitable that we will end up missing some and I do think this code will likely add a non trivial overhead to datafusion)
> 
> >

Yes, I agree. So I think we may check the depth of exprs using `safe_recursion` in `datafusion/src/sql/planner.rs`, then we can avoid using `may_growth` in the follow-up process.

**Comment by xudong963** [comment]
> > > This is an acceptable short term workaround, but I think it would be more efficient and more elegant if we rewrite these recursive procedures into iterative procedures.
> 
> > Do you mean to use push-based model？
> 
> @xudong963 , I think what @houqp is suggesting is to rewrite code that is recursive to not be recursive.
> 
> The pattern for Datafusion probably looks like taking code like
> 
> ```rust
> fn visit(expr: &expr)  {
>   for child in expr.children() {
>     visit(child)
>   }
>   // do actual expr logic
> }
> ```
> 
> And changing it so the state is tracked with a structure on the heap rather than a stack. I think [`VecDeque`](https://doc.rust-lang.org/stable/std/collections/struct.VecDeque.html) is a good one for rust:
> 
> ```rust
> fn visit(expr: &expr)  {
>   let mut worklist = VecDequeue::new();
>   worklist.push_back(expr);
>   while !worklist.is_empty() {
>     let parent = worklist.pop_front();
>     for child in parent.children() {
>       worklist.push_back(child)
>     }
>     // do actual expr logic on parent
>   }
> }
> ```
> 
> (aka avoid the call back to `visit`)

Open an issue https://github.com/apache/arrow-datafusion/issues/1444 to track of this.

**Comment by xudong963** [comment]
I'll continue to hack on the ticket on the weekend 💪

**Comment by alamb** [comment]
Marking as stale pr -- will close it in a week or two unless we plan to keep working on it

**Comment by xudong963** [comment]
> Marking as stale pr -- will close it in a week or two unless we plan to keep working on it

I'll directly close the ticket because I plan to fix it by #1444

**Pull request closed without merging**